### ISSUE TITLE
[FEATURE] - or Question? Auto start VPN if leaving the trusted SSID

### ISSUE THREAD

**Issue by vsatmydynipnet** [zaneschepke/wgtunnel#60]
Hi,

first, thank you for this cool peace of software. I am moving from openvpn to wireguard and have installed your app from fdroid.

My current settings with OpenVPN are:

- When I leave my home SSID Easer (automation app)  starts OpenVPN to my home and sets Owntrack in move mode
- when coming home Easer stops OpenVPN and sets Owntracks in pause mode

Now i try to have a similiar setting with wgtunnel, but i think i oversee something.

- i set "tunnel on mobile data" to on, everything else to off and added my SSID as trusted SSID
- tried it without SSID in trusted too

if i switch the peer on, it tunnels no matter i am on WLAN or LTE.  If I switch the tunnel off it does not tunnel even leaving my SSID.

It is probably me not seeing something here.

If this does currently not work, is there a command where i can switch the peer on and off with Easer?

Many thanks for any hint!

**Comment by zaneschepke** [comment]
Hello! So the whole first block of settings on the settings screen is for auto tunneling settings. It sounds like you were configuring things correctly, you just need to "Enable auto-tunneling" to actually make those settings active. I'm sorry that this screen is a bit confusing. I'm thinking about changing it to be more intuitive. You will likely want to enable battery saver as well, or else the service that monitors the network changes will drain the battery.

**Comment by vsatmydynipnet** [comment]
Hi thank you ,for your answer. I tried that and have locatiion services enabled always - se ttings does not show the screen anymore - but when trying to enable auto tunneling if fails with - location services required - 

Lineageos 20, rooted with magisk, no gapps

Location service works for home assistant, organic maps, owntracks, so i assume device setup is ok

**Comment by zaneschepke** [comment]
> Hi thank you ,for your answer. I tried that and have locatiion services enabled always - se ttings does not show the screen anymore - but when trying to enable auto tunneling if fails with - location services required - 
> 
> Lineageos 20, rooted with magisk, no gapps
> 
> Location service works for home assistant, organic maps, owntracks, so i assume device setup is ok

Can you confirm you have "allow all the time" location permission enabled and "precise location" enabled?

**Comment by vsatmydynipnet** [comment]
from mobile i write from pc after uplosd

![Screenshot_20231126-191133_WG Tunnel](https://github.com/zaneschepke/wgtunnel/assets/18069867/0a1c469d-c1a9-423d-91cf-8978ca81e29b)
![Screenshot_20231126-191113_Einstellungen](https://github.com/zaneschepke/wgtunnel/assets/18069867/11830aa1-d49a-4b9a-ba1b-0891e125ce60)
![Screenshot_20231126-191034_Berechtigungssteuerung](https://github.com/zaneschepke/wgtunnel/assets/18069867/d14a2806-a11c-4c8f-aa68-9ac7207e9351)

**Comment by vsatmydynipnet** [comment]
I am sorry its in German,  but these are the related screenshots. Its always and precise, but wgtunnel does not request the locations as shown in the other screenshot. homeassistant and others access it.

**Comment by zaneschepke** [comment]
Thank you for making me aware of this. I think there must be something wrong with how the app is checking for location services on LOS. I will put LOS on one of my devices and see if I can find the issue.

**Comment by vsatmydynipnet** [comment]
Thank you for looking into this. Let me know if I can be of any help.

**Comment by bastei** [comment]
For me this works on LOS 20. But location services must not only be granted to the app, they actually have to be enabled in the Android settings. Otherwise auto-tunneling can't be toggled (yes, you can't even deactivate it) – you only get the "Location services required" message.
As location services are just needed for the trusted wifi filter to work, maybe the app should just show a warning when location services are disabled instead of preventing whole auto-tunneling to get activated/deactivated, @zaneschepke?

**Comment by zaneschepke** [comment]
> For me this works on LOS 20. But location service permissions must not only be granted to the app, they actually have to be enabled in the Android settings. Otherwise auto-tunneling can't be toggled (yes, you can't even deactivate it) – you only get the "Location services required" message. As location services are just needed for the trusted wifi filter to work, maybe the app should just show a warning when location services are disabled instead of preventing whole auto-tunneling to get activated/deactivated, @zaneschepke?

Thank you for this info @bastei! @vsatmydynipnet  Can you try and enable location services in Android settings? 

As for your idea, I think this would be good for disabling auto-tunneling. I will add that capability. Permissions should not be prevent disabling auto-tunneling. I still want to prevent auto-tunneling enablement without (location services + tunneling untrusted wifi) because it will simply not work. It could then cause reports/issues of it not working that will be hard to diagnose. 

I will add that it is possible to enable auto-tunneling without location services if you are not using it for tunneling on wifi networks and are just using it for tunneling on mobile data.

**Comment by bastei** [comment]
> I still want to prevent auto-tunneling enablement without (location services + tunneling untrusted wifi) because it will simply not work. It could then cause reports/issues of it not working that will be hard to diagnose.

In my opinion that is not enough. What about users that disable location services in Android settings _after_ enabling auto-tunneling with (trusted) wifi? Without any hint, the wifi filter stops working then, which is unexpected and you will get reports/issues again. ;)
I think that this scenario is not that rare. Users temporarily disable location services and sometimes they even forget to enable it again. As auto-tunnel is a core feature of wgtunnel and users rely on that feature, they better get informed about unexpected situations.

I think the following steps would help usability-wise:
1. Add a hint to the "Add trusted wifi name" input, for example: "Disabling the tunnel for trusted wifi connections only works while location services are enabled in the Android settings. The tunnel stays active otherwise and you have to disable the tunnel manually for trusted wifi connections."
Optionally add a status info (for example "Location services are currently enabled" + green light / "Locations services are currently disabled" + red light).
2. While auto-tunneling (with wifi) is enabled, listen for [MODE_CHANGED_ACTION](https://developer.android.com/reference/android/location/LocationManager#MODE_CHANGED_ACTION). When location services get disabled, inform the user on every wifi connection with a notification that wgtunnel enables the tunnel regardless of entered trusted wifis because it can't access wifi information with location services disabled.
3. Add an option "Disable the tunnel for the current wifi connection" to the notification, so that the user can easily disable the tunnel if the current wifi is actually trusted. Enable the tunnel again when [onCapabilitiesChanged](https://github.com/zaneschepke/wgtunnel/blob/16e65aec9fdd06f35bb1d5c6e0410f4aa492d7bb/app/src/main/java/com/zaneschepke/wireguardautotunnel/service/network/BaseNetworkService.kt#L39) gets called.

This way, users stay informed about the current situation and are enabled to correct the behaviour of wgtunnel. They can even suppress the notifications if they don't fit to their personal user-case.

**Comment by zaneschepke** [comment]
I think these are all good suggestions. I have a pretty full backlog at the moment, but perhaps some time will open up to implement this in the future. Thank you for such a detailed response and thoughtful approach to the problem!

**Comment by Trinnik** [comment]
I want to jump in here because I had an issue on LOS where I could not enable auto-tunnel if "Tunnel on untrusted wifi" was enabled at all. I ensured location was enabled all of the time, precise location was checked, battery optimization was turned off, the whole nine yards. The location services page in Android settings was blank and the quick tile appeared as though Location Services were enabled.

In the end, what fixed it for me was using the quick tile to disable location and re-enable it. That triggered some other popup about location services terms. Once accepting that and ensuring location was turned back on I was able to enable auto-tunneling with untrusted wifi enabled.

Hopefully that's helpful to someone else who comes across this issue. Seemed to be some weird LOS bug where location services weren't really enabled even though everything appeared as though they were.
